src/lib/datatypes/sol-memdesc.c

```diff
@@ -0,0 +1,804 @@
+{
+    const struct sol_vector *v = memory;
+
+    errno = ERANGE;
+    SOL_INT_CHECK(idx, > UINT16_MAX, NULL);
+    errno = 0;
+
+    return sol_vector_get(v, idx);
+}
+
+static int
+vector_ops_resize_array(const struct sol_memdesc *array, void *memory, size_t len)
+{
+    struct sol_vector *v = memory;
+    uint16_t oldlen;
+
+    SOL_INT_CHECK(len, > UINT16_MAX, -ERANGE);
+
+    oldlen = v->len;
+    if (oldlen == len)
+        return 0;
+
+    if (oldlen < len) {
+        void *m = sol_vector_append_n(v, len - oldlen);
+
+        if (!m)
+            return errno ? -errno : -ENOMEM;
+
+        if (sol_memdesc_get_size(array->array_item)) {
+            uint16_t idx;
+
+            for (idx = oldlen; idx < len; idx++) {
+                void *itmem;
+                int r;
+
+                itmem = sol_vector_get_nocheck(v, idx);
+                r = sol_memdesc_init_defaults(array->array_item, itmem);
+                if (r < 0) {
+                    sol_vector_del_range(v, idx, len - idx);
+                    return r;
+                }
+            }
+        }
+
+        return 0;
+    } else {
+        if (sol_memdesc_get_size(array->array_item)) {
+            uint16_t idx;
+
+            for (idx = len; idx < oldlen; idx++) {
+                void *itmem;
+
+                itmem = sol_vector_get_nocheck(v, idx);
+                sol_memdesc_free_content(array->array_item, itmem);
+            }
+        }
+
+        return sol_vector_del_range(v, len, oldlen - len);
+    }
+}
+
+SOL_API const struct sol_memdesc_ops SOL_MEMDESC_OPS_VECTOR = {
+    SOL_SET_API_VERSION(.api_version = SOL_MEMDESC_OPS_API_VERSION, )
+    .init_defaults = vector_ops_init_defaults,
+    .array = &(const struct sol_memdesc_ops_array){
+        SOL_SET_API_VERSION(.api_version = SOL_MEMDESC_OPS_ARRAY_API_VERSION, )
+        .get_length = vector_ops_get_array_length,
+        .get_element = vector_ops_get_array_element,
+        .resize = vector_ops_resize_array,
+    },
+};
+
+
+static int
+ptr_vector_ops_init_defaults(const struct sol_memdesc *desc, void *mem)
+{
+    struct sol_ptr_vector *v = mem;
+
+    if (desc->size != sizeof(struct sol_ptr_vector))
+        return -EINVAL;
+
+    if (desc->array_item && sol_memdesc_get_size(desc->array_item) != sizeof(void *))
+        return -EINVAL;
+
+    sol_ptr_vector_init(v);
+
+    if (desc->defcontent.p)
+        return sol_memdesc_set_content(desc, mem, desc->defcontent.p);
+
+    return 0;
+}
+
+SOL_API const struct sol_memdesc_ops SOL_MEMDESC_OPS_PTR_VECTOR = {
+    SOL_SET_API_VERSION(.api_version = SOL_MEMDESC_OPS_API_VERSION, )
+    .init_defaults = ptr_vector_ops_init_defaults,
+    .array = &(const struct sol_memdesc_ops_array){
+        SOL_SET_API_VERSION(.api_version = SOL_MEMDESC_OPS_ARRAY_API_VERSION, )
+        .get_length = vector_ops_get_array_length,
+        .get_element = vector_ops_get_array_element,
+        .resize = vector_ops_resize_array,
+    },
+};
```

Comment:
I thought about this as well, likewise why serialize is here and not in sol-buffer.h

The reasoning I've come with is that these headers are much more used than sol-memdesc.h and forcing the dependency/inclusion will offer no good. Then these constants are provided here.

Also, if the API ever changes, these will be in the same source file and it's easier to fix.